In [1]:
#  Enable HTML/CSS 
from IPython.core.display import HTML
HTML("<link href='https://fonts.googleapis.com/css?family=Passion+One' rel='stylesheet' type='text/css'><style>div.attn { font-family: 'Helvetica Neue'; font-size: 30px; line-height: 40px; color: #FFFFFF; text-align: center; margin: 30px 0; border-width: 10px 0; border-style: solid; border-color: #5AAAAA; padding: 30px 0; background-color: #DDDDFF; }hr { border: 0; background-color: #ffffff; border-top: 1px solid black; }hr.major { border-top: 10px solid #5AAA5A; }hr.minor { border: none; background-color: #ffffff; border-top: 5px dotted #CC3333; }div.bubble { width: 65%; padding: 20px; background: #DDDDDD; border-radius: 15px; margin: 0 auto; font-style: italic; color: #f00; }em { color: #AAA; }div.c1{visibility:hidden;margin:0;height:0;}div.note{color:red;}</style>")

___
Enter Team Member Names here (double click to edit):

- Name 1:
- Name 2:
- Name 3:

________

# In Class Assignment Three
In the following assignment you will be asked to fill in python code and derivations for a number of different problems. Please read all instructions carefully and turn in the rendered notebook (or HTML of the rendered notebook)  before the end of class.

<a id="top"></a>
## Contents
* <a href="#Loading">Loading the Data</a>
* <a href="#distance">Measuring Distances</a>

** Available after class begins: **
* <a href="#KNN">K-Nearest Neighbors</a>
* <a href="#naive">Naive Bayes</a>

________________________________________________________________________________________________________
<a id="Loading"></a>
<a href="#top">Back to Top</a>
## Downloading the Document Data
Please run the following code to read in the "20 newsgroups" dataset from sklearn's data loading module.

In [2]:
from sklearn.datasets import fetch_20newsgroups_vectorized
import numpy as np
#from __future__ import print_function

# this takes about 30 seconds to compute, read the next section while this downloads
ds = fetch_20newsgroups_vectorized(subset='train')

# this holds the continuous feature data (which is tfidf)
print('features shape:', ds.data.shape) # there are ~11000 instances and ~130k features per instance
print('target shape:', ds.target.shape) 
print('range of target:', np.min(ds.target),np.max(ds.target))
print('Data type is', type(ds.data), float(ds.data.nnz)/(ds.data.shape[0]*ds.data.shape[1])*100, '% of the data is non-zero')

features shape: (11314, 130107)
target shape: (11314,)
range of target: 0 19
Data type is <class 'scipy.sparse.csr.csr_matrix'> 0.1214353154362896 % of the data is non-zero


## Understanding the Dataset
Look at the description for the 20 newsgroups dataset at http://qwone.com/~jason/20Newsgroups/. You have just downloaded the "vectorized" version of the dataset, which means all the words inside the articles have gone through a transformation that binned them into 130 thousand features related to the words in them.  

**Question Set 1**:
- How many instances are in the dataset? 
- What does each instance represent? 
- How many classes are in the dataset and what does each class represent?
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?
- Is the data represented as a sparse or dense matrix?

___
Enter your answer here:

* There are 11314 instances in the dataset
* Each instance represents a newgroup article
* There are 20 classes ranging form 0 to 19. Each class represents a type of article - rangin from topics form science, receration, computers, politics, religion, etc.
* It depends on the population of articles that we are looking at in the past week. If they belong to the same category, then it might work, but if they belong to different categoriers that were not included in the training data, then the results would be misleading.
* This data is represented as a sparse matrix since < 0.125% of the array contains any non-sero information.

___
<a id="distance"></a>
<a href="#top">Back to Top</a>
## Measures of Distance
In the following block of code, we isolate three instances from the dataset. The instance "`a`" is from the group *computer graphics*, "`b`" is from from the group *recreation autos*, and "`c`" is from group *recreation motorcycle*. **Exercise for part 2**: Calculate the: 
- (1) Euclidean distance
- (2) Cosine distance 
- (3) Jaccard similarity 

between each pair of instances using the imported functions below. Remember that the Jaccard similarity is only for binary valued vectors, so convert vectors to binary using a threshold. 

**Question for part 2**: Which distance seems more appropriate to use for this data? **Why**?

In [28]:
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import jaccard
import numpy as np

# get first instance (comp)
idx = 550
a = ds.data[idx].todense()
a_class = ds.target_names[ds.target[idx]]
print('Instance A is from class', a_class)

# get second instance (autos)
idx = 4000
b = ds.data[idx].todense()
b_class = ds.target_names[ds.target[idx]]
print('Instance B is from class', b_class)

# get third instance (motorcycle)
idx = 7000
c = ds.data[idx].todense()
c_class = ds.target_names[ds.target[idx]]
print('Instance C is from class', c_class)

# Enter distance comparison below for each pair of vectors:
p = 'Placeholder'

euc_ab = euclidean(a,b)
euc_ac = euclidean(a,c)
euc_bc = euclidean(b,c)

cos_ab = cosine(a,b)
cos_ac = cosine(a,c)
cos_bc = cosine(b,c)

# For Jaccard
# Vector must be boolean
a_discrete = a 
a_discrete[a_discrete > 0] = 1
#print(a_discrete.sum())

b_discrete = b 
b_discrete[b_discrete > 0] = 1
#print(b_discrete.sum())

c_discrete = c 
c_discrete[c_discrete > 0] = 1
#print(c_discrete.sum())

jac_ab = jaccard(a_discrete,b_discrete)
jac_ac = jaccard(a_discrete,c_discrete)
jac_bc = jaccard(b_discrete,c_discrete)

print('\n\nEuclidean Distance\n ab:', euc_ab, 'ac:', euc_ac, 'bc:',euc_bc)
print('Cosine Distance\n ab:', cos_ab, 'ac:', cos_ac, 'bc:', cos_bc)
print('Jaccard Dissimilarity (vectors should be boolean values)\n ab:', jac_ab, 'ac:', jac_ac, 'bc:', jac_bc)

print('\n\nThe most appropriate distance is...')
print("Cosine")

print ("\nEucledian is not appropriate since the data is so sparse that most features will be similar to one another between instances (most likely 0's)")
print ("\nJaccard may be appropriate to some extent since we dont really care about 0's, we care about the fields that have data. \
But since we have artificially set a threshold to discretize, it may introduce some issues. \
More importantly though it is just looking at what fields are similar between the vectors. \
2 vectors may have the same number of similar fields but the actual fields that are similar may be different fron one another. \
In such a case, Jaccard distances may be close to one another even though the actual vectors may not have any relationship to each other. \
This can likely be seen for the jaccard distance ab and bc - no relationship between these but distances are similar in value")
print ("\nCosine mat be appropriate since it is looking at angles between vectors. Similar instance will be pointing in the same direction and hence this might be the best metric. \
Also, the data is sparse and cosine distances work well on sparse data")

Instance A is from class comp.graphics
Instance B is from class rec.autos
Instance C is from class rec.motorcycles


Euclidean Distance
 ab: 1.0985184671870858 ac: 1.1891405425398234 bc: 0.9177794226661624
Cosine Distance
 ab: 0.6033714113755322 ac: 0.7070276149559529 bc: 0.4211595343347173
Jaccard Dissimilarity (vectors should be boolean values)
 ab: 0.8821138211382114 ac: 0.8754716981132076 bc: 0.9087947882736156


The most appropriate distance is...
Cosine

Eucledian is not appropriate since the data is so sparse that most features will be similar to one another between instances (most likely 0's)

Jaccard may be appropriate to some extent since we dont really care about 0's, we care about the fields that have data. But since we have artificially set a threshold to discretize, it may introduce some issues. More importantly though it is just looking at what fields are similar between the vectors. 2 vectors may have the same number of similar fields but the actual fields that are simi